rfcs/convergence/lockstep-versioning.md

```diff
@@ -0,0 +1,145 @@
+
+## Detailed Design or Proposal
+
+### Rollup the react-components changelog and remove the version bump entries
+
+Work has already been done in [#19349](https://github.com/microsoft/fluentui/pull/19349), but additional work is
+required to add this kind of functionality into our CI release flow. This work **will** be completed regardless of
+the result of the other proposals in this RFC, since the `@fluentui/react-components` changelog currently provides
+no meaningful information to users.
+
+### Slower release cadence
+
+A slower release cadence will allow us to:
+
+- Plan work, PRs and validation before release.
+- Release a coherent set of changes with clear migration plan for version upgrades.
+- Reduce the number of version bump entries in changelogs that don't provide a lot of information.
+- Set clear expectations for partners when they can expect to consume expected changes/fixes. There should be very
+  little reason that partners can't plan around a slower release cadence.
+
+This RFC proposes a **monthly** automated release cadence for Fluent v9 packages. There should be a clear way to
+run ad-hoc releases when necessary, to ship urgent features and hotfixes.
+
+Daily/Nightly releases be published under a `9.x.x-nightly.commitSha` prerelease tag, for anyone that wants to
+test features from the bleeding edge.
+
+### Lockstep all v9 packages
+
+All v9 packages should be bumped in lockstep. Any release bumps every package with the maximum change
+type in the version group (major, minor, patch). Below is an example of how it would work:
+
+```
+// Before
+@fluentui/react-components 9.0.0
+@fluentui/react-button     9.0.0 -> patch
+@fluentui/react-accordion  9.0.0 -> patch
+@fluentui/react-utilities  9.0.0 -> minor
+
+
+// After
+@fluentui/react-components 9.1.0
+@fluentui/react-button     9.1.0
+@fluentui/react-accordion  9.1.0
+@fluentui/react-utilities  9.1.0
+```
+
+Once partners install `react-components` as a single suite package they can investigate versioning issues
+but checking if every v9 dependency is on the same version as the suite package. Installing separate component/utility
+packages also becomes easy, since partners should just make sure that the resolved versions are all the same.
+
+Our versioning and release tool `beachball` currently has a version group feature, which can support lockstep if all
+packages are already on the same version. This feature was does not actually guarantee lockstep, we should continue work
+on it so that it supports lockstep versioning strictly.
+
+Lockstep versioning is only feasible with the slower release cadence. A daily release cadence would mean the same
+version problem we currently have with `react-components` but for every single v9 package.
+
+### Pros and Cons
+
+#### Pros
+
+- Easier debuggability of dependency tree
+- More coherent changelogs for the suite package
+
+#### Cons
+
+- Packages can be bumped with no changes
```

Comment:
I call this "phantom version bumping"; when you increase the version of a thing for no reason other than "easier debuggability of the dep tree". This is the single largest contributor to bundle bloat in app partners.

Consider this:

App partners will not "flip a switch" and convert to v9+. They will be in an interim state, possibly perpetually, between versions. Even if they did update everything, there is no guarantee their dependencies will do the same. We are seeing this today with v7 => v8, where various partners have different cadences up updates. Some partners have restrictions that prevent them from updating too fast, like the cadence of their downstream dependencies conflicting, causing them to take the most conservative approach.

Given that you can not control cadences downstream, the best possible approach is to prioritize what's good for the apps: dedupe any and all dependencies possible. This is better than optimizing for debuggability of dependencies, or servicing within the repo.

Take for example the `utilities` or `merge-styles` packages in v7 or v8. They have not major changed in perhaps 4 years, yet they have gone through multiple major releases due to lock step majors. Yet, when partners who may be on v8 take a dependency on something that's in v7, they end up with 2 copies of these packages in the bundle. This was unnecessary and there is just nothing they can do to fix it other than to bloat things. Now multiply this by all the code. There is literally > 300k of duplicate fluent deps in office apps right now because of this.

This can be avoided if packages are treated like individual entities with strong version contracts. If you avoid phantom version bumping, you can still major bump parts of your code graph, reducing the duplication to only what's absolutely required. Components are where we want to major release more frequently, because things like behavioral changes or css changes are more common in there. Foundational things like css utilities and hooks and such should almost never change. This means when partners go from v9 to v10 and inevitably include both in the graph, the duplication is minimized and the components from the two versions can share dependencies. It also means that scenarios like: "v10 only major bumped the slider" means you don't have 2 copies of the unchanged DetailsList which is a travesty that's happening today.

If things are versioned correctly, there should be no reason why one component should not work with another library assuming semvers match and they are respected. React is not lock step. Popper isn't. TypeScript isn't. Shoot, even the fluent web-components aren't lock stepping. How do we ever get along without everything lock stepping? Because they all follow semver requirements. So I think a better question is how we enforce semver better; are there ways to associate api surface movement to change files? Should shifts in the styling be associated with major bumps? What kind of problems are we hitting with "version mismatches" and what automation should we have in place to work through those?

Think of it this way; if every package in the repo was in their own repos, how would we ensure they all work well together? This is how we should approach this concern in order for partner apps to avoid being penalized, because they and their deps need to be able to move at their own cadences, which really means our only options in making that work are to optimize for our repo needs and cause them to download dupes, or optimize for them and ensure our packages work well together whether they live in one repo or in separate repos.

Reply:
Another thought. Suite packages (the `react-components` package which exports all the others) needs to "follow" versioning of its dependencies, since it re-exports the contract. That is: if you major bump the `react-tooltip` dependency, you will need to major bump the `react-components` version itself, since the contract of the suite is now changing in a major way.

I don't think this is enforced today by beachball. I think a package needs to have some metadata available to indicate that it re-exports another package, so that this can be automatic.

The suite package is a way to simplify the dependency management and import story for partners, without creating phantom version bumps. I also love your point on changelogs being useful - often they are not, and I agree that releasing too frequently in an automated way, without any manual scrub of the content, is a contributor to poor changelog content.

I think another idea to improve changelog content is to beef up the change editing. When a dev gets that prompt from beachball saying "Enter a description of the change", they often have no idea their comments are going to be part of the documentation. I think a better editing experience, complete with examples of what "good comments" look like, would help improve what developers type.

Reply:
Thanks for the detailed post. From what I understand the big bundlesize headaches come when a major bump happens and causes parters to duplicate more dependencies than they need to.

Following independent semver bumping would mean that only non-compatible versions are duplicated ?

I'll make sure to reflect this in the RFC

Reply:
I cannot agree more with @dzearing, our application have as much as 7 fabric versions duplicated in the bundle at one point. And that is simply due to the fact that indirect dependencies (our dependencies that depends on fabric) causing it.

I would like to add that unnecessary major bumps not only cause bundle size issues and slow npm installs due to duplications, we also observed they are causing:
  1. Unnecessary and hard to resolve type errors due to duplications. In the form of `import('node_modules/office-ui-fabric-react_v7.165.1/libs/Button.tsx').IButtonProps is not assignable to import('node_modules/office-ui-fabric-react_v7.170.2/libs/Button.tsx').IButtonProps` when this happens, there's very little our developers can do apart from casting.
  2. Significant slow down in TypeScript and Webpack compilation of the application. After a lot of cross-team effort to remove the fabric duplications, we have observed the compilation speed up by 20%.
  3. Unsatisfiable peer dependencies will be more frequent with more major versions. Where app have two dependencies in the package.json each demanding a different major version of fabric as peer.
  4. Strange bugs happen where we get different theme in different part of the app.

I appreciate that fabric 8 break up the monolithic package and go with the smaller packages. That resolves an important issue for us where we want to take in a bug fix on one part of fabric but is often blocked by other part of the monolithic fabric package. Now that it is broken up, the information of "which package is compatible with which other package" needs to be stored somewhere, and that is in the version ranges of your package.json. From app's perspective, we really hope that fabric can get more accurate on the compatibility range in package.json and further break down packages to reduce unnecessary dependencies.

Reply:
> 7 fabric versions duplicated in the bundle at one point

 @weichensw  While I agree with the major bumps causing duplication issues, 7 versions either means that you have all 7 versions of Fabric (1-7) in your dep chain or semver is not resolving dependencies correctly. Are patch and minor versions duplicated for you ? because they shouldn't be and that is not  a problem with Fluent versions.

Reply:
@ling1726 Thanks for the response. Sorry I should have clarified on this. Our application is not built with a single package.json but multiple package.jsons. So when we eventually linking them together, multiple minor versions can exist even if we deduped each one of them properly. Also our direct dependencies might not have a wide range for fabric versions, some of them pin to a minor version of fabric also is another source of duplication.

So the cause of the duplication is currently not related to "too many major versions" (and we only had 7 major versions so far). But if we have a lot of major bumps, then we already know it will cause more duplications. I am giving above examples to add that duplication not only cause bundle size (performance) issues which we normally consider, but also lead to other (correctness) issues as mentioned above.

Reply:
> Our application is not built with a single package.json but multiple package.jsons. So when we eventually linking them together, multiple minor versions can exist even if we deduped each one of them properly

@weichensw I think I understand what you are saying here and  normally this would scream alarm bells for me ⚠️⚠️

I don't think any versioning scheme that Fluent follows could resolve this issue for you in either the versioning scheme proposed in this RFC (lockstep) versus versioning each package independently in Fluent which is what @dzearing is talking about, since you cannot accurately rely on semver to actually deduplicate packages with your package manager.

**If anything versioning packages independently might actually make this problem worse for you**.  in Lockstep you would have the following version scheme:

* A - x.x.x
* B - x.x.x
* C - x.x.x

If any Fluent packages deviate from that single version then you know that you have deduping/resolving work to do 💪💪 and it's instantly clear that there is a problem.

In this the below case, no matter how we declare which packages are compatible with which, how would you know which to version to start from ? especially now that major version bumps for one Fluent package no longer have any meaning for other Fluent packages

* A - x.y.z
* B - 1.2.3
* C  - a.b.c

> But if we have a lot of major bumps

This seems interesting, what is `a lot of major bumps` for you ? If Fluent as a whole major bumped every 18 months is that considered too much ?

Reply:
@weichensw @ling1726 Maybe we should talk internally sometime about the details of your setup and concerns. Can you message me on Teams? (Looks like there are multiple MS people with your name, so not sure who to ping. On that note, if you could link your account at https://repos.opensource.microsoft.com/people that would be helpful.)

Reply:
Hi @ling1726, 
> This seems interesting, what is a lot of major bumps for you ? If Fluent as a whole major bumped every 18 months is that considered too much ?

Sorry I meant to say "if we have a lot of unnecessary bumps".

Like I mentioned earlier. The duplicated minors or indirect dependencies pinning on minor version is not in anyway fabric's fault and does not need to be resolved by fabric's versioning. What I meant is that (because due to those we were suffered a lot by duplicated versions, which now we kind of resolved them) we have fairly good experience on the _issues caused by duplicated versions_. I meant to back up David by saying that normally we view duplicated versions as a bundle-size problem only, but in practice it does also cause a lot of other issues that might not be obvious like bundle size impacts.

So I think doing some extra work to ensure the compatibility range of packages is accurate worth it, considering the problems duplicated versions can cause (is more troublesome than it appears to be).

Reply:
Trying to follow the nuances here, but also want to call out:

Consumers of FUI should be instructed to use caret deps. Especially true for middle-layers like LPC and admin components. One thing I caught recently was a middle layer using exact version dependencies. This is a recipe for introducing 7 copies of fluent ui.

We can be certain that dupes WILL happen at the app layer, because all consumers in an app graph don't upgrade their deps at the same time.

With that said, can we agree to NOT lockstep major any more? This will help with deduping a *majority* of dependencies in the fui package graph which never major bump, while leaving others which do contain major changes to be duplicated until the full graph of partners upgrade.

We should also recommend partners use caret dependencies (not peer deps.) This gives the widest range that's "compatible" while also allowing for duplicates to deal with out-of-sync upgrades in the graph. Exact and tilde dependencies should be discouraged as the allowed version range is minimized and duplication likelihood is maximized.

Side note on peers: Peer deps (as referenced in the MUI docs and in comments elsewhere) has this problem: only the app host declares a single winner. What happens when LPC requires fui 9 peer and OOUI requires fui 10 peer? App host has to stop bumping until the lower dependency is resolved, or they have to do unnatural things to force duplication. There are cases like "react" where stopping the upgrade train while versions get in sync makes sense. But I think fui is something where 2 versions might be desirable; e.g. ooui has a flag to "turn on v10" while v9 is the default, so they end up both living in the graph.

Reply:
@dzearing this is a good point around peers. @weichensw and my team have talked about going to deps to address those issues. We currently list FUI as a peer with caret but it creates the issue you mentioned where you might have to unnaturally force duplication. We are also hoping to get even more granular to allow bundlers to resolve the graph better and as you said reduce the number of bumps. Our theme and util packages we try to keep as stable as possible while breaking and duping a component package isn't nearly as big of a deal since it's at the end of the dep chain. Appreciate all the discussion here guys. In admin at least I think we lean towards not lock stepping for all the reasons mentioned above so just wanted to +1 those thoughts.